diff --git a/src/contrib/fairscheduler/src/java/org/apache/hadoop/mapred/FairSchedulerServlet.java b/src/contrib/fairscheduler/src/java/org/apache/hadoop/mapred/FairSchedulerServlet.java
index f562ebd..5bb18f4 100644
--- src/contrib/fairscheduler/src/java/org/apache/hadoop/mapred/FairSchedulerServlet.java
+++ src/contrib/fairscheduler/src/java/org/apache/hadoop/mapred/FairSchedulerServlet.java
@@ -128,12 +128,23 @@ public class FairSchedulerServlet extends HttpServlet {
     out.printf("<title>%s Job Scheduler Admininstration</title>\n", hostname);
     out.print("<link rel=\"stylesheet\" type=\"text/css\" " + 
         "href=\"/static/hadoop.css\">\n");
+    out.print("<link rel=\"stylesheet\" type=\"text/css\" " + 
+        "href=\"/static/style.css\">\n");
+	out.print("<style>#wrapper { margin: 0px !important; padding: 0px !important}</style>");
     out.print("</head><body>\n");
-    out.printf("<h1><a href=\"/jobtracker.jsp\">%s</a> " + 
+	out.print("<div id='wrapper'><div id='title'>");
+    out.printf("<h1>%s " + 
         "Job Scheduler Administration</h1>\n", hostname);
+	out.print("</div>");
+	out.print("<div class='section'>");
     showPools(out, advancedView);
+	out.print("</div>");
+	out.print("<div class='section'>");
     showJobs(out, advancedView);
+	out.print("</div>");
+	out.print("<div class='section'>");
     showAdminForm(out, advancedView);
+	out.print("</div>");
     out.print("</body></html>\n");
     out.close();
   }
diff --git a/src/webapps/job/jobtracker.jsp b/src/webapps/job/jobtracker.jsp
index 73a0a98..1ec2743 100644
--- src/webapps/job/jobtracker.jsp
+++ src/webapps/job/jobtracker.jsp
@@ -96,9 +96,9 @@
 <link rel="stylesheet" type="text/css" href="/static/style.css">
 </head>
 <body>
-
 <div id="wrapper">
   <div id="title">
+	<a href="http://www.cloudera.com/community-support" target="suport" class="support">Need help? Visit Cloudera Community Support</a>
 <h1><%= trackerName %> Hadoop Map/Reduce Administration</h1>
     <div id="status">
 
@@ -150,8 +150,18 @@
 Job Tracker History</a>
 
 </div>
+
+<%  if (tracker.infoServer.getAttribute("scheduler") != null) {
+    %>
+<iframe src="/scheduler" style="width:961px; height: 500px; border: none;"></iframe>
+<%
+  } else {
+    %><p>(FairScheduler is not running.)</p><%
+  }
+%>
+
 </div>
 
 <%
-out.println(ServletUtil.htmlFooter());
+//out.println(ServletUtil.htmlFooter());
 %>
\ No newline at end of file
diff --git a/src/webapps/static/style.css b/src/webapps/static/style.css
index d8b71a0..e823d54 100644
--- src/webapps/static/style.css
+++ src/webapps/static/style.css
@@ -20,6 +20,12 @@ a.more {
 	background:url(/static/images/more_arrow.png) no-repeat right center;
 	padding-right:8px;
 }
+a.support {
+	display: block;
+	float: right;
+	position: relative;
+	top: 2px;
+}
 
 div.detail {
 	color:#999;
