Index: src/core/org/apache/hadoop/fs/s3/S3Exception.java
===================================================================
--- src/core/org/apache/hadoop/fs/s3/S3Exception.java	(revision 761128)
+++ src/core/org/apache/hadoop/fs/s3/S3Exception.java	(working copy)
@@ -18,10 +18,12 @@
 
 package org.apache.hadoop.fs.s3;
 
+import java.io.IOException;
+
 /**
  * Thrown if there is a problem communicating with Amazon S3.
  */
-public class S3Exception extends RuntimeException {
+public class S3Exception extends IOException {
 
   public S3Exception(Throwable t) {
     super(t);
