diff --git src/examples/pipes/Makefile.am src/examples/pipes/Makefile.am
index 731ab1e..a28e4a6 100644
--- src/examples/pipes/Makefile.am
+++ src/examples/pipes/Makefile.am
@@ -17,7 +17,7 @@ ACLOCAL_AMFLAGS = -I ../../c++/utils/m4
 AM_CXXFLAGS=-Wall -I$(HADOOP_UTILS_PREFIX)/include \
             -I$(HADOOP_PIPES_PREFIX)/include
 LDADD=-L$(HADOOP_UTILS_PREFIX)/lib -L$(HADOOP_PIPES_PREFIX)/lib \
-      -lhadooppipes -lhadooputils
+      -lhadooppipes -lhadooputils -lpthread
 
 bin_PROGRAMS= wordcount-simple wordcount-part wordcount-nopipe pipes-sort
 
diff --git src/examples/pipes/Makefile.in src/examples/pipes/Makefile.in
index 100270e..b8462f7 100644
--- src/examples/pipes/Makefile.in
+++ src/examples/pipes/Makefile.in
@@ -217,7 +217,7 @@ AM_CXXFLAGS = -Wall -I$(HADOOP_UTILS_PREFIX)/include \
             -I$(HADOOP_PIPES_PREFIX)/include
 
 LDADD = -L$(HADOOP_UTILS_PREFIX)/lib -L$(HADOOP_PIPES_PREFIX)/lib \
-      -lhadooppipes -lhadooputils
+      -lhadooppipes -lhadooputils -lpthread
 
 
 # Define the sources for each program
